File: core/src/main/java/dev/bee/kanjianki/core/StudyTaskCopy.java
package dev.bee.kanjianki.core;

public final class StudyTaskCopy {
    private static final String LABEL_STUDY = "Study";
    private static final String LABEL_SIMILAR_KANJI = "Similar kanji";

    private static final String TASK_TARGETED_FLASHCARD = "targeted_flashcard";
    private static final String TASK_MEANING_FLASHCARD = "meaning_flashcard";
    private static final String TASK_FONT_RECOGNITION = "font_recognition";
    private static final String TASK_REPAIR_WRITING = "repair_writing";
    private static final String TASK_TARGETED_WRITING = "targeted_writing";
    private static final String TASK_CONTEXT_WRITING = "context_writing";
    private static final String TASK_GUIDED_WRITING = "guided_writing";
    private static final String TASK_BLIND_WRITING = "blind_writing";
    private static final String TASK_SAMPLED_HANDWRITING = "sampled_handwriting";
    private static final String TASK_CONFUSABLE_RECOGNITION = "confusable_recognition";

    private StudyTaskCopy() {
    }

    public static String labelForTask(String task) {
        if (task == null) {
            return LABEL_STUDY;
        }
        return switch (task) {
            case TASK_TARGETED_FLASHCARD -> "Focused recall";
            case StudyTaskTypes.KANJI_MEANING -> "Kanji -> meaning";
            case StudyTaskTypes.MEANING_KANJI -> "Meaning -> kanji";
            case StudyTaskTypes.TYPING_MEANING, StudyTaskTypes.TYPE_MEANING -> "Type the meaning";
            case StudyTaskTypes.FONT_MEANING -> "Font -> meaning";
            case StudyTaskTypes.WORD_READING -> "Word -> reading";
            case StudyTaskTypes.WRITE_KANJI -> "Write kanji";
            case StudyTaskTypes.SIMILAR_KANJI -> LABEL_SIMILAR_KANJI;
            case TASK_MEANING_FLASHCARD -> "Quick recall";
            case TASK_FONT_RECOGNITION -> "Font check";
            case TASK_REPAIR_WRITING -> "Write to repair";
            case TASK_TARGETED_WRITING -> "Focused practice";
            case TASK_CONTEXT_WRITING -> "New problem kanji";
            case TASK_GUIDED_WRITING -> "Guided review";
            case TASK_BLIND_WRITING, TASK_SAMPLED_HANDWRITING -> "Memory check";
            case TASK_CONFUSABLE_RECOGNITION -> "Learn the shape";
            default -> LABEL_STUDY;
        };
    }

    public static String flashcardTitle(RecordsSchedulerModels.StudySession session) {
        if (isWordReadingTask(session)) {
            return "Read this word";
        }
        if (isTypingMeaningTask(session)) {
            return "Type the meaning";
        }
        if (isMeaningKanjiTask(session)) {
            return "Choose the kanji";
        }
        return isFontRecognitionTask(session) ? "Recognise this kanji" : "Name this kanji";
    }

    public static String studyModeLabel(RecordsSchedulerModels.StudySession session) {
        if (session != null && session.writingRequired) {
            return "Practice";
        }
        if (isWordReadingTask(session)) {
            return "Read";
        }
        if (isTypingMeaningTask(session)) {
            return "Type";
        }
        if (isMeaningKanjiTask(session)) {
            return "Recall";
        }
        return "Recognise";
    }

    public static boolean isTeachingTask(RecordsSchedulerModels.StudySession session) {
        if (session == null) {
            return false;
        }
        return TASK_CONTEXT_WRITING.equals(session.taskType)
                || TASK_GUIDED_WRITING.equals(session.taskType)
                || (TASK_TARGETED_WRITING.equals(session.taskType) && session.item.learningStep < 2);
    }

    public static boolean isRecallTask(RecordsSchedulerModels.StudySession session) {
        if (session == null) {
            return false;
        }
        return TASK_BLIND_WRITING.equals(session.taskType) || TASK_SAMPLED_HANDWRITING.equals(session.taskType);
    }

    public static boolean isFontRecognitionTask(RecordsSchedulerModels.StudySession session) {
        return session != null
                && (StudyTaskTypes.FONT_MEANING.equals(session.taskType)
                || TASK_FONT_RECOGNITION.equals(session.taskType));
    }

    public static boolean isTypingMeaningTask(RecordsSchedulerModels.StudySession session) {
        return session != null
                && (StudyTaskTypes.TYPING_MEANING.equals(session.taskType)
                || StudyTaskTypes.TYPE_MEANING.equals(session.taskType));
    }

    public static boolean isMeaningKanjiTask(RecordsSchedulerModels.StudySession session) {
        return session != null && StudyTaskTypes.MEANING_KANJI.equals(session.taskType);
    }

    public static boolean isWordReadingTask(RecordsSchedulerModels.StudySession session) {
        return session != null && StudyTaskTypes.WORD_READING.equals(session.taskType);
    }
}

File: core/src/main/kotlin/dev/bee/kanjianki/core/StudyTaskCopy.kt
package dev.bee.kanjianki.core

import java.util.Locale

object StudyTaskCopy {
    private const val JAPANESE_LANGUAGE = "ja"
    private const val LABEL_STUDY = "Study"
    private const val LABEL_SIMILAR_KANJI = "Similar kanji"

    private const val TASK_TARGETED_FLASHCARD = "targeted_flashcard"
    private const val TASK_MEANING_FLASHCARD = "meaning_flashcard"
    private const val TASK_FONT_RECOGNITION = "font_recognition"
    private const val TASK_CONFUSABLE_RECOGNITION = "confusable_recognition"

    @JvmStatic
    fun labelForTask(task: String?): String = when (task) {
        null -> localizedText(LABEL_STUDY, "学習")
        TASK_TARGETED_FLASHCARD -> localizedText("Focused recall", "集中復習")
        StudyTaskTypes.KANJI_MEANING -> localizedText("Kanji -> meaning", "漢字→意味")
        StudyTaskTypes.MEANING_KANJI -> localizedText("Meaning -> kanji", "意味→漢字")
        StudyTaskTypes.TYPING_MEANING,
        StudyTaskTypes.TYPE_MEANING -> localizedText("Type the meaning", "意味を入力")

        StudyTaskTypes.FONT_MEANING -> localizedText("Font -> meaning", "フォント→意味")
        StudyTaskTypes.WORD_READING -> localizedText("Word -> reading", "単語→読み")
        StudyTaskTypes.TYPE_READING -> localizedText("Type the reading", "読みを入力")
        StudyTaskTypes.KANJI_READING -> localizedText("Kanji -> reading", "漢字→読み")
        StudyTaskTypes.READING_KANJI -> localizedText("Reading -> kanji", "読み→漢字")
        StudyTaskTypes.SENTENCE_READING -> localizedText("Sentence reading", "文で読む")
        StudyTaskTypes.WRITE_KANJI -> localizedText("Write kanji", "漢字を書く")
        StudyTaskTypes.SIMILAR_KANJI -> localizedText(LABEL_SIMILAR_KANJI, "似た漢字")
        TASK_MEANING_FLASHCARD -> localizedText("Quick recall", "素早く復習")
        TASK_FONT_RECOGNITION -> localizedText("Font check", "フォント確認")
        StudyTaskTypes.REPAIR_WRITING -> localizedText("Repair", "修正")
        StudyTaskTypes.TARGETED_WRITING -> localizedText("Focused practice", "集中練習")
        StudyTaskTypes.CONTEXT_WRITING -> localizedText("New problem kanji", "新しい問題漢字")
        StudyTaskTypes.GUIDED_WRITING -> localizedText("Guided review", "ガイド付き復習")
        StudyTaskTypes.BLIND_WRITING,
        StudyTaskTypes.SAMPLED_HANDWRITING -> localizedText("Memory check", "記憶確認")

        TASK_CONFUSABLE_RECOGNITION -> localizedText("Learn the shape", "形を覚える")
        else -> localizedText(LABEL_STUDY, "学習")
    }

    @JvmStatic
    fun studyModeLabel(session: RecordsSchedulerModels.StudySession?): String = when {
        isNewLearningRepeat(session) -> localizedText("Learn", "学習")
        isRelearning(session) -> localizedText("Relearning", "再学習")
        session != null && session.writingRequired -> localizedText("Practice", "練習")
        isTypingReadingTask(session) -> localizedText("Type", "入力")
        isSentenceReadingTask(session) -> localizedText("Read", "読む")
        isWordReadingTask(session) -> localizedText("Read", "読む")
        isTypingMeaningTask(session) -> localizedText("Type", "入力")
        isKanjiReadingTask(session) -> localizedText("Choose", "選ぶ")
        isReadingKanjiTask(session) -> localizedText("Choose", "選ぶ")
        isMeaningKanjiTask(session) -> localizedText("Recall", "思い出す")
        else -> localizedText("Recognise", "見分ける")
    }

    private fun isNewLearningRepeat(session: RecordsSchedulerModels.StudySession?): Boolean {
        val item = session?.item ?: return false
        return item.phase == RecordsBase.SchedulerPhase.NEW_LEARNING && item.totalReviews > 0
    }

    private fun isRelearning(session: RecordsSchedulerModels.StudySession?): Boolean =
        session?.item?.phase == RecordsBase.SchedulerPhase.RELEARNING

    @JvmStatic
    fun isTeachingTask(session: RecordsSchedulerModels.StudySession?): Boolean {
        if (session == null) {
            return false
        }
        return StudyTaskTypes.CONTEXT_WRITING == session.taskType ||
            StudyTaskTypes.GUIDED_WRITING == session.taskType ||
            (
                StudyTaskTypes.TARGETED_WRITING == session.taskType &&
                    (session.item?.learningStep ?: Int.MAX_VALUE) < 2
                )
    }

    @JvmStatic
    fun isRecallTask(session: RecordsSchedulerModels.StudySession?): Boolean {
        if (session == null) {
            return false
        }
        return StudyTaskTypes.BLIND_WRITING == session.taskType ||
            StudyTaskTypes.SAMPLED_HANDWRITING == session.taskType
    }

    @JvmStatic
    fun isRepairWritingTask(session: RecordsSchedulerModels.StudySession?): Boolean =
        session != null && StudyTaskTypes.REPAIR_WRITING == session.taskType

    @JvmStatic
    fun isFontRecognitionTask(session: RecordsSchedulerModels.StudySession?): Boolean =
        session != null &&
            (StudyTaskTypes.FONT_MEANING == session.taskType || TASK_FONT_RECOGNITION == session.taskType)

    @JvmStatic
    fun isTypingMeaningTask(session: RecordsSchedulerModels.StudySession?): Boolean =
        session != null &&
            (StudyTaskTypes.TYPING_MEANING == session.taskType || StudyTaskTypes.TYPE_MEANING == session.taskType)

    @JvmStatic
    fun isTypingReadingTask(session: RecordsSchedulerModels.StudySession?): Boolean =
        session != null && StudyTaskTypes.TYPE_READING == session.taskType

    @JvmStatic
    fun isMeaningKanjiTask(session: RecordsSchedulerModels.StudySession?): Boolean =
        session != null && StudyTaskTypes.MEANING_KANJI == session.taskType

    @JvmStatic
    fun isWordReadingTask(session: RecordsSchedulerModels.StudySession?): Boolean =
        session != null &&
            (StudyTaskTypes.WORD_READING == session.taskType || StudyTaskTypes.TYPE_READING == session.taskType)

    @JvmStatic
    fun isKanjiReadingTask(session: RecordsSchedulerModels.StudySession?): Boolean =
        session != null && StudyTaskTypes.KANJI_READING == session.taskType

    @JvmStatic
    fun isReadingKanjiTask(session: RecordsSchedulerModels.StudySession?): Boolean =
        session != null && StudyTaskTypes.READING_KANJI == session.taskType

    @JvmStatic
    fun isSentenceReadingTask(session: RecordsSchedulerModels.StudySession?): Boolean =
        session != null && StudyTaskTypes.SENTENCE_READING == session.taskType

    private fun localizedText(english: String, japanese: String): String =
        if (isJapaneseLocale()) japanese else english

    private fun isJapaneseLocale(): Boolean = Locale.getDefault().language == JAPANESE_LANGUAGE
}
